### PR TITLE
Unify search boxes

### PR DESCRIPTION
**Pull request by denyskon**
Unify all but a few search boxes to use uniform style, uniform translations and shared templates where possible.
Remove a few duplicated search templates, e. g. code search.

<details><summary>Example after screenshots:</summary>

grafik

grafik

grafik

grafik

</details>

Also includes #29700 

Co-authored-by: 6543 <6543@obermui.de>

### CODE REVIEW

File: options/locale/locale_en-US.ini [hunk 1]
```diff
@@ -1989,17 +1986,6 @@ contributors.contribution_type.commits = Commits
...
-search = Search
-search.search_repo = Search repository
-search.type.tooltip = Search type
-search.fuzzy = Fuzzy
-search.fuzzy.tooltip = Include results that also matches the search term closely
-search.match = Match
-search.match.tooltip = Include only results that matches the exact search term
-search.results = Search results for "%s" in <a href="%s">%s</a>
-search.code_no_results = No source code matching your search term found.
-search.code_search_unavailable = Currently code search is not available. Please contact your site administrator.
```
Comment (denyskon) [inline options/locale/locale_en-US.ini:2047]: How did that even happen....? Exact duplicate of line 340+ :laughing:

File: options/locale/locale_en-US.ini [hunk 2]
```diff
@@ -142,6 +142,19 @@ confirm_delete_selected = Confirm to delete all selected items?
...
 
+[search]
+search = Search
+search_kind = Search %s...
+type_tooltip = Search type
+fuzzy = Fuzzy
+fuzzy_tooltip = Include results that also match the search term closely
+match = Match
+match_tooltip = Include only results that match the exact search term
+no_results = No %s matching your search term found.
```
Comment (delvh) [inline options/locale/locale_en-US.ini]: Not a good idea to deduplicate this term.
Or are you using it as `ctx.Locale.Tr("…no_results", ctx.Locale.Tr("…repo"))` for example?

File: templates/repo/search.tmpl [hunk 3]
```diff
@@ -20,12 +20,9 @@
 		</div>
 		{{if .CodeIndexerUnavailable}}
 			<div class="ui error message">
-				<p>{{ctx.Locale.Tr "repo.search.code_search_unavailable"}}</p>
+				<p>{{ctx.Locale.Tr "search.unavailable" (ctx.Locale.Tr "search.kind.code")}}</p>
 			</div>
 		{{else if .Keyword}}
-			<h3>
-				{{ctx.Locale.Tr "repo.search.results" .Keyword .RepoLink .RepoName}}
-			</h3>
```
Comment (delvh) [inline templates/repo/search.tmpl:28]: ?

File: options/locale/locale_en-US.ini (hunk 2 above)
Comment (delvh) [reply options/locale/locale_en-US.ini]: Yep, you did exactly that…

**Review by delvh — DISMISSED** [review]

File: templates/repo/search.tmpl (hunk 3 above)
Comment (denyskon) [reply templates/repo/search.tmpl:28]: I don't see a reason for this to be there. In every other case we simply output the search results, but with code search we first print this header containing the search term (already visible in the search field above) and the repo (should be obvious)

File: options/locale/locale_en-US.ini (hunk 2 above)
Comment (lafriks) [reply options/locale/locale_en-US.ini]: This is very tricky, while it works in English it will not work in other languages (word casing, word forms etc, for example in Latvian each noun has 7 forms for each singular and multiple cases with different endings depending on the sentence where it's used), so explore labels definitely cannot be used here, it's definitely better to duplicate such strings than concatenate them as it makes translations harder or even impossible

Just for example English word Repository will have Latvian translation - `repozitorijs` (and 6 additional forms - repozitorija, repozitirijam, repozitorijā, repozitoriju, ar repozitoriju, repozitorij) and for Repositories will have additional 6 forms - repozitoriji, repozitorijiem, repozitorijos, repozitoriju, repozitorijus, ar repozitorijiem... So please don't reuse nouns from labels in sentences

So in Latvian explore label it would be "**Repozitoriji**" while in this sentence "Netika atrasts neviens **repozitorijs**, kas atbilstu kritērijiem" (as you can see word form and first letter case changes as in Latvian language you can not write upper case letters in words in middle of sentence, only for special words like person names, countries etc) and for "Search Repositories..." translation would be "Meklēt **repozitorijus**..." (even different form)

**Review by lafriks — DISMISSED** [review]
As per my comment

File: options/locale/locale_en-US.ini [hunk 4]
```diff
@@ -142,6 +142,19 @@ confirm_delete_selected = Confirm to delete all selected items?
 name = Name
 value = Value
 
+[search]
+search = Search
+search_kind = Search %s...
```
Comment (lafriks) [inline options/locale/locale_en-US.ini]: Same here

File: options/locale/locale_en-US.ini (hunk 2 above)
Comment (denyskon) [reply options/locale/locale_en-US.ini]: @lafriks Maybe simply do "Search..." everywhere where it's obvious?

**Comment by denyskon** [comment]
I will refactor search templates as a whole.....

**Comment by fnetX** [comment]
Hi there, I'm just checking by … I am currently also refactoring some search boxes for Forgejo for accessibility, and my work is probably exactly the opposite of this pull. I cannot block your decisions, but please consider the following.

> Maybe simply do "Search..." everywhere where it's obvious?

The meaning of search boxes is usually obvious to people with intact eye vision, depending on where a search box is placed. A search box at the top of the page conveys a global search, a search box just on top of a repo sidebar conveys it searches the repos. However, if the visual layout of the page is not available (and the semantic representation does not help either – mostly the case in this codebase), a screenreader reading "Search" to you leaves you in the dark about what you are about to search.

The simplest solution would be to add a "Search x" string for every search box including an aria-label or labelled-by element, because a placeholder on its own does not satisfy accessibility requirements. If you really want, you can keep the Search ... placeholder, but in any case I recommend to keep the possibility to specify the exact meaning of a search box.

And in localization, there is also often good reason to have duplicate terms, because grammar rules differ in many languages. Reusing strings is often done at the level of localization tools itself, and while I cannot say anything about the proprietary Crowdin, at least Weblate has a "translation memory" which offers translators access to similar strings – still keeping the ability to adapt if needed.

File: options/locale/locale_en-US.ini (hunk 2 above)
Comment (lafriks) [reply options/locale/locale_en-US.ini]: I would not mind from my perspective but see @fnetX comment and I agree that it's better to have specific placeholder for accessibility

**Comment by denyskon** [comment]
@fnetX @lafriks @delvh

I now refactored the whole bunch of search templates, including using shared templates where possible. I also added a corresponding & uniform "Search *... (repos/orgs/users/*" translation key for most search boxes to ensure a11y.

File: templates/shared/search/input.tmpl [hunk 5]
```diff
@@ -0,0 +1,4 @@
+<!-- Value - value of the search field (for search results page) -->
+<!-- Disabled (optional) - if search field has to be disabled -->
+<!-- Placeholder (optional) - placeholder text to be used -->
+<input type="search" spellcheck="false" name="q" maxlength="255" placeholder="{{if .Placeholder}}{{.Placeholder}}{{else}}{{ctx.Locale.Tr "search.search"}}…{{end}}"{{if .Value}} value="{{.Value}}"{{end}}{{if .Disabled}} disabled{{end}} autofocus>
```
Comment (6543) [inline templates/shared/search/input.tmpl]: if disabled 'disable' **else** 'autofocus'

**Review by 6543 — APPROVED** [review]

**Comment by silverwind** [comment]
I like unified style, but I'm not sure I like primary color on the search button. It calls too much attention imho for a secondary action.

File: templates/shared/search/button.tmpl [hunk 6]
```diff
@@ -0,0 +1,3 @@
+<!-- Disable (optional) - if search button has to be disabled -->
+<!-- Tooltip (optional) - a tooltip to be displayed on hover -->
+<button class="ui small primary icon button" aria-label="{{ctx.Locale.Tr "search.search"}}" {{with .Tooltip}}data-tooltip-content="{{.}}"{{end}}{{if .Disabled}} disabled{{end}}>{{svg "octicon-search"}}</button>
```
Comment (silverwind) [inline templates/shared/search/button.tmpl]: What does `with` do and shouldn't that be an `if`?

Comment (denyskon) [reply templates/shared/search/button.tmpl]: If `.Tooltip` is set, the following code block will be executed and `.` becomes the value of `.Tooltip`. It's basically just a cleaner notation compared to the if-one and I use it pretty often. If you don't like it in Gitea I can also change it to if.

Comment (silverwind) [reply templates/shared/search/button.tmpl]: I guess `if` would be more common in gitea code. Changing the value of `.` is also often a problem as seen in loops, but I guess in this simple case it's ok.

Comment (denyskon) [reply templates/shared/search/button.tmpl]: I would like to keep `with` if it's not a blocker for you, as this one-line templates should stay as minimalistic as possible and this notation is simply shorter and easier to understand from my pov

**Comment by denyskon** [comment]
I think our UI is getting more and more monochrome and it isn't a good thing. We already removed the color from the whole repo header, let's at least keep colored buttons..... I think the primary color is exactly right there

File: templates/shared/search/button.tmpl (hunk 6 above)
Comment (silverwind) [reply templates/shared/search/button.tmpl]: I forbid `with` in my JS, but I guess for go template it might be borderline ok 😆

Comment (6543) [reply templates/shared/search/button.tmpl]: I also had to look twice but i find it more elegant.

And if with is used more often people will notice imidiatly :)

**Comment by 6543** [comment]
As it is a refactor pull ... if this is the only issue to address i would argue to revert it as @silverwind suggested and after merge propose that change in a second pull

**Comment by denyskon** [comment]
grafik

Okay, I've reverted the primary color. But we really should do something against the greyscalification of our UI :laughing:

**Comment by silverwind** [comment]
> But we really should do something against the greyscalification of our UI 😆

I see it as a evolution of the design. Take a look at other forges like GitLab or GitHub, there is almost no color left. I think it does give a bit more "professional" look.

**Comment by 6543** [comment]
We should not discusst this out here ... but i personally woul find more collor beter to navigate etc ...

**Comment by denyskon** [comment]
@silverwind Do you have any other concerns?

**Comment by silverwind** [comment]
I have not reviewed in detail yet. But I recall @wxiaoguang had some reservations about this PR.

**Comment by wxiaoguang** [comment]
> @wxiaoguang What exactly do you think is over-engineering here? :)

1. Personally I do not like the "{{if}}{{else}}" trick in templates.
2. This PR mixes too many different things together: "fuzzy" change, moved code, changed variable, changed logic
3. A lot of untested code, for example, at first glance, why `{{template "shared/search/input" dict "Value" .Query (ctx.Locale.Tr "search.package_kind")}}`? The "dict" accept key-value pairs, but I could only see 3 arguments there.

File: templates/package/shared/list.tmpl [hunk 7]
```diff
@@ -2,15 +2,15 @@
 {{if .HasPackages}}
 <form class="ui form ignore-dirty">
 	<div class="ui fluid action input">
-		{{template "shared/searchinput" dict "Value" .Query}}
-		<select class="ui dropdown" name="type">
+		{{template "shared/search/input" dict "Value" .Query (ctx.Locale.Tr "search.package_kind")}}
```
Comment (denyskon) [inline templates/package/shared/list.tmpl]:
```suggestion
		{{template "shared/search/input" dict "Value" .Query "Placeholder" (ctx.Locale.Tr "search.package_kind")}}
```

**Comment by silverwind** [comment]
There is a height mismatch on /milestones page because of missing `small` class on `ui fluid action input`. `/issues` and `pulls` are okay but please check all occurences.

image

**Comment by silverwind** [comment]
Can something be done to make all search bars 35.4px? On `/explore` pages, it's 40px for example. Adding `small` there does not reduce the height.

File: templates/explore/user_list.tmpl [hunk 8]
```diff
@@ -26,6 +26,12 @@
...
 		</div>
 	{{else}}
-		<div class="flex-item">{{ctx.Locale.Tr "explore.user_no_results"}}</div>
+		<div class="flex-item">
+		{{if .PageIsExploreUsers}}
+			{{ctx.Locale.Tr "search.user_no_results"}}
+		{{else}}
+			{{ctx.Locale.Tr "search.org_no_results"}}
+		{{end}}
+		</div>
```
Comment (wxiaoguang) [inline templates/explore/user_list.tmpl]: Could it be a general "search.no_results" ?

File: templates/repo/branch/list.tmpl [hunk 9]
```diff
@@ -67,140 +67,136 @@
 			</div>
 		{{end}}
 
-		{{if .Branches}}
```
Comment (wxiaoguang) [inline templates/repo/branch/list.tmpl:70]: The removed line seems to be a unrelated change. If I understand correctly, this PR's purpose is to "unify search boxes", but not "change the branch render logic" ....

File: templates/repo/commits_table.tmpl [hunk 10]
```diff
@@ -8,27 +8,30 @@
...
 			<a href="{{$.CommitRepoLink}}/commit/{{.BeforeCommitID | PathEscape}}" class="ui green sha label gt-mx-0">{{if not .BaseIsCommit}}{{if .BaseIsBranch}}{{svg "octicon-git-branch"}}{{else if .BaseIsTag}}{{svg "octicon-tag"}}{{end}}{{.BaseBranch}}{{else}}{{ShortSha .BaseBranch}}{{end}}</a>
 			...
 			<a href="{{$.CommitRepoLink}}/commit/{{.AfterCommitID | PathEscape}}" class="ui green sha label gt-mx-0">{{if not .HeadIsCommit}}{{if .HeadIsBranch}}{{svg "octicon-git-branch"}}{{else if .HeadIsTag}}{{svg "octicon-tag"}}{{end}}{{.HeadBranch}}{{else}}{{ShortSha .HeadBranch}}{{end}}</a>
-		{{end}}
-	</div>
+		</div>
+	{{end}}
 </h4>
 
+{{if .PageIsCommits}}
```
Comment (wxiaoguang) [inline templates/repo/commits_table.tmpl:20]: The old UI is done intentionally, to make it compact. The new UI makes the "input" a full line? Is it really good (and important) on the commit list page?

image

File: templates/repo/home.tmpl [hunk 11]
```diff
@@ -15,14 +15,12 @@
...
-								<input name="q" value="{{.Keyword}}"{{if .CodeIndexerUnavailable}} disabled{{end}} placeholder="{{ctx.Locale.Tr "repo.search.search_repo"}}">
+							<div class="ui small action input{{if .CodeIndexerUnavailable}} disabled left icon{{end}}"{{if .CodeIndexerUnavailable}} data-tooltip-content="{{ctx.Locale.Tr "search.code_search_unavailable"}}"{{end}}>
+								<input name="q" value="{{.Keyword}}"{{if .CodeIndexerUnavailable}} disabled{{end}} placeholder="{{ctx.Locale.Tr "search.code_kind"}}">
 								{{if .CodeIndexerUnavailable}}
 									<i class="icon">{{svg "octicon-alert"}}</i>
 								{{end}}
-								<button class="ui small icon button"{{if .CodeIndexerUnavailable}} disabled{{end}} type="submit">
-									{{svg "octicon-search"}}
-								</button>
+								{{template "shared/search/button"}}
```
Comment (wxiaoguang) [inline templates/repo/home.tmpl]: Missing `{{if .CodeIndexerUnavailable}}`?

File: templates/shared/repo_search.tmpl [hunk 12]
```diff
@@ -65,3 +65,4 @@
 		<span data-tooltip-content="{{ctx.Locale.Tr "explore.relevant_repositories_tooltip"}}">{{ctx.Locale.Tr "explore.relevant_repositories" (printf "?only_show_relevant=0&sort=%s&q=%s&language=%s" $.SortType (QueryEscape $.Keyword) (QueryEscape $.Language))}}</span>
 	</div>
 {{end}}
+<div class="divider"></div>
```
Comment (wxiaoguang) [inline templates/shared/repo_search.tmpl:68]: Why an extra "divider" here .... maybe it is done to make the UI look better, but maybe it could also not be considered as the purpose of "unifying search boxes"

File: templates/shared/search/button.tmpl [hunk 13]
```diff
@@ -0,0 +1,3 @@
+<!-- Disable (optional) - if search button has to be disabled -->
+<!-- Tooltip (optional) - a tooltip to be displayed on hover -->
```
Comment (wxiaoguang) [inline templates/shared/search/button.tmpl]: Could it use Golang template comment syntax? (here and others)

File: templates/shared/search/code/results.tmpl [hunk 14]
```diff
@@ -0,0 +1,44 @@
+<div class="flex-text-block gt-fw">
+	{{$pageURL := print $.SourcePath "/search"}}
+	{{if and $.PageIsExploreCode $.ContextUser}}
+		{{$pageURL = print AppSubUrl "/" $.ContextUser.Name "/-/code"}}
+	{{else if $.PageIsExploreCode}}
+		{{$pageURL = print AppSubUrl "/explore/code"}}
+	{{end}}
```
Comment (wxiaoguang) [inline templates/shared/search/code/results.tmpl]: That's really tricky. These "if" blocks might be broken in the future and difficult to test.

Could it make the caller prepare the "PageURL" ahead?

File: templates/shared/search/code/results.tmpl [hunk 15]
```diff
@@ -0,0 +1,44 @@
...
+		<i class="color-icon gt-mr-3" style="background-color: {{$term.Color}}"></i>
+		{{$term.Language}}
+		<div class="detail">{{$term.Count}}</div>
+	</a>
+	{{end}}
+</div>
+<div class="repository search">
+	{{range $result := .SearchResults}}
+		{{$repoLink := $.SourcePath}}
+		{{if $.PageIsExploreCode}}{{$repoLink = (index $.RepoMaps .RepoID).Link}}{{end}}
```
Comment (wxiaoguang) [inline templates/shared/search/code/results.tmpl]: It also looks fragile here and below (by using `if`), that's what I meant "over-engineering" .....

File: templates/shared/search/input.tmpl [hunk 16]
```diff
@@ -0,0 +1,4 @@
+<!-- Value - value of the search field (for search results page) -->
+<!-- Disabled (optional) - if search field has to be disabled -->
+<!-- Placeholder (optional) - placeholder text to be used -->
+<input type="search" spellcheck="false" name="q" maxlength="255" placeholder="{{if .Placeholder}}{{.Placeholder}}{{else}}{{ctx.Locale.Tr "search.search"}}…{{end}}"{{if .Value}} value="{{.Value}}"{{end}}{{if .Disabled}} disabled{{else}} autofocus{{end}}>
```
Comment (wxiaoguang) [inline templates/shared/search/input.tmpl]: Should "autofocus" always be applied? I think in some cases the search input shouldn't grab the focus ahead.

Comment (wxiaoguang) [inline templates/shared/search/input.tmpl]: Maybe the dots in `{{ctx.Locale.Tr "search.search"}}…` should be also put into the translation string, to make it translatable, just like "repo_kind", etc

**Review by wxiaoguang — COMMENTED** [review]
A brief review without really testing. 

TBH, I would suggest one PR should only focus on major purpose (unless mixing something is not avoidable, or the changes are small enough and all in one module/package)

Mixing too many things together makes it difficult to review or test, and would cause conflicts with many PRs frequently.

https://github.com/go-gitea/gitea/blob/main/CONTRIBUTING.md#pull-request-format

image

File: options/locale/locale_en-US.ini [hunk 17]
```diff
@@ -155,6 +155,30 @@ filter.not_template = Not Template
...
+repo_kind = Search repos...
+repo_no_results = No matching repositories found.
+user_kind = Search users...
+user_no_results = No matching users found.
+org_kind = Search orgs...
+org_no_results = No matching organizations found.
+team_kind = Search teams...
+code_kind = Search code...
+code_no_results = No matching source code found.
+code_search_unavailable = Code search is currently not available. Please contact your site administrator.
```
Comment (wxiaoguang) [inline options/locale/locale_en-US.ini]: For a large instance, "your" seems not a proper word.

For example, if GitHub crashes, GitHub users couldn't find "their" site administrator. 🤣 


So maybe it could be "Please contact the site administrator."

File: templates/repo/branch/list.tmpl (hunk 9 above)
Comment (denyskon) [reply templates/repo/branch/list.tmpl:70]: The reason is that otherwise the search bar disappears when there are no results....

File: templates/repo/commits_table.tmpl (hunk 10 above)
Comment (denyskon) [reply templates/repo/commits_table.tmpl:20]: It's the way it is done on other tables, e. g. on all the admin pages. I think it is better so stay consistent, and I don't see a problem in one more line used.

File: templates/repo/home.tmpl (hunk 11 above)
Comment (denyskon) [inline templates/repo/home.tmpl]:
```suggestion
								{{template "shared/search/button" dict "Disabled" .CodeIndexerUnavailable}}
```

File: templates/shared/repo_search.tmpl (hunk 12 above)
Comment (denyskon) [reply templates/shared/repo_search.tmpl:68]: It is present on the other pages like explore/users or explore/orgs. I could have moved it out to a separate PR, but I don't think it's necessary.

File: templates/shared/search/code/results.tmpl (hunk 15 above)
Comment (wxiaoguang) [reply templates/shared/search/code/results.tmpl]: If there are so many different details, I think it's better to use 2 separate template files, instead of mixing the "if" logics into one template file.

File: templates/shared/search/input.tmpl (hunk 16 above)
Comment (silverwind) [reply templates/shared/search/input.tmpl]: Imho, we should just remove all these autofocus. They annoy me very much and also are bad for accessibilty because they break things like spacebar scrolling. GitHub also does not do this.

File: options/locale/locale_en-US.ini (hunk 17 above)
Comment (silverwind) [reply options/locale/locale_en-US.ini]: Or just remove the second sentence entirely.

File: templates/shared/search/input.tmpl (hunk 16 above)
Comment (silverwind) [reply templates/shared/search/input.tmpl]: Also see [this](https://developer.mozilla.org/en-US/docs/Web/HTML/Global_attributes/autofocus#accessibility_concerns):

> Automatically focusing a form control can confuse visually-impaired people using screen-reading technology and people with cognitive impairments. When autofocus is assigned, screen-readers "teleport" their user to the form control without warning them beforehand.

I think autofocus is only justified in a few cases like opening a new form to focus the first element or in cases like google where search is the primary action.

File: templates/repo/commits_table.tmpl (hunk 10 above)
Comment (silverwind) [reply templates/repo/commits_table.tmpl:20]: Could move the search to the empty space above the table in this case.

File: templates/explore/user_list.tmpl (hunk 8 above)
Comment (silverwind) [reply templates/explore/user_list.tmpl]: Or just `no_results` even. We need to cut down on unnecessary translation namespacing.

File: options/locale/locale_en-US.ini (hunk 17 above)
Comment (delvh) [inline options/locale/locale_en-US.ini]:
```suggestion
code_search_unavailable = Code search is currently unavailable.
```

File: options/locale/locale_en-US.ini [hunk 18]
```diff
@@ -155,6 +155,30 @@ filter.not_template = Not Template
...
+package_kind = Search packages...
+project_kind = Search projects...
+branch_kind = Search branches...
+commit_kind = Search commits...
+runner_kind = Search runners...
+keyword_search_unavailable = Searching by keyword is currently not available. Please contact your site administrator.
```
Comment (delvh) [inline options/locale/locale_en-US.ini]:
```suggestion
keyword_search_unavailable = Keyword search is currently unavailable.
```

File: templates/admin/org/list.tmpl [hunk 19]
```diff
@@ -7,7 +7,26 @@
 			</div>
 		</h4>
 		<div class="ui attached segment">
-			{{template "admin/base/search" .}}
```
Comment (delvh) [inline templates/admin/org/list.tmpl:10]: Why did you delete that template?
Why not simply leave it in?

File: templates/shared/search/button.tmpl [hunk 20]
```diff
@@ -0,0 +1,3 @@
...
+<button class="ui small icon button" aria-label="{{ctx.Locale.Tr "search.search"}}" {{with .Tooltip}}data-tooltip-content="{{.}}"{{end}}{{if .Disabled}} disabled{{end}}>{{svg "octicon-search"}}</button>
```
Comment (delvh) [inline templates/shared/search/button.tmpl:3]:
```suggestion
<button class="ui small icon button" aria-label="{{ctx.Locale.Tr "search.search"}}" {{if .Tooltip}}data-tooltip-content title="{{.Tooltip}}"{{end}} {{if .Disabled}} disabled{{end}}>{{svg "octicon-search"}}</button>
```

File: templates/shared/search/code/results.tmpl (hunk 14 above)
Comment (delvh) [reply templates/shared/search/code/results.tmpl]: The parent template, or the Go code?

File: templates/shared/search/input.tmpl (hunk 16 above)
Comment (denyskon) [reply templates/shared/search/input.tmpl]: Removed

File: templates/repo/commits_table.tmpl (hunk 10 above)
Comment (denyskon) [reply templates/repo/commits_table.tmpl:20]: @silverwind I tried now, however I didn't manage to fix the styles involved :laughing:  I would leave it as done currently. This commits table should be reworked somewhen soon, and I can probably try to do it if you don't have time to continue your PR.

**Comment by denyskon** [comment]
@silverwind The height should be okay now. It's a mix of fomantic paddings (where do these magic numbers come from :cry: ) and missing "small" class....

File: templates/shared/search/code/results.tmpl (hunk 14 above)
Comment (denyskon) [reply templates/shared/search/code/results.tmpl]: I will check again what can be done with this template....

File: templates/repo/commits_table.tmpl (hunk 10 above)
Comment (silverwind) [reply templates/repo/commits_table.tmpl:20]: Feel free to, I don't think I will be motivated for that PR anytime soon 😆

File: web_src/css/form.css [hunk 21]
```diff
@@ -41,6 +41,11 @@ textarea,
   color: var(--color-input-text);
 }
 
+/* fix fomantic small dropdown having inconsistent padding with input */
+.ui.small.selection.dropdown {
+  padding: .67857143em 3.2em .67857143em 1em
```
Comment (silverwind) [inline web_src/css/form.css]:
```suggestion
  padding: .67857143em 3.2em .67857143em 1em;
```

**Comment by silverwind** [comment]
Ok, I will test heights again. Yeah I know it's tricky. We should eventually just get rid of fomantic classes for search altogether, which will hopefully be easier now that so much is in shared templates.

**Comment by silverwind** [comment]
Found tiny issue: Borders are not continous here. Checking for a fix...

image

**Comment by denyskon** [comment]
The problem isn't really search but dropdowns in all their variety. Ideally we would make an own dropdown module

**Comment by silverwind** [comment]
Some CSS fixed done. This was a bit tricky.

Screenshot 2024-03-14 at 20 40 40

Screenshot 2024-03-14 at 20 40 28
Screenshot 2024-03-14 at 20 40 22

**Comment by silverwind** [comment]
Ideally we would find a suitable framework with headless components that we can style our own, like these, but they are React/Vue only unfortunately:

- https://www.radix-ui.com/
- https://headlessui.com/

I imagine there must be similar frameworks out there for Vanilla JS.

File: templates/repo/commits_table.tmpl [hunk 22]
```diff
@@ -8,27 +8,30 @@
...
+	<div class="ui attached segment">
+		<form class="ignore-dirty" action="{{.RepoLink}}/commits/{{.BranchNameSubURL}}/search">
+			<div class="ui small fluid action input">
+				{{template "shared/search/input" dict "Value" .Keyword "Placeholder" (ctx.Locale.Tr "search.commit_kind")}}
+				<select class="ui small dropdown" name="all">
+					<option value="false">{{ctx.Locale.Tr "repo.commits.search_branch"}}</option>
+					<option {{if .All}} selected="selected"{{end}} value="true">{{ctx.Locale.Tr "repo.commits.search_all"}}</option>
+				</select>
```
Comment (silverwind) [inline templates/repo/commits_table.tmpl]: Can you extract this branch dropdown to a file? Should be very similar to `fuzzy.tmpl`. I noticed that fomantic-ui transforms the `<select>` to `<div>` during rendering, but if we pre-render it correctly, that transformation shouldn't be needed.

Reason I noticed was because the `dropdown icon` was not SVG, but icon font.

**Review by silverwind — APPROVED** [review]
https://github.com/go-gitea/gitea/pull/29530#discussion_r1525418103 would be nice, otherwise I'm happy.

File: templates/shared/search/code/results.tmpl (hunk 14 above)
Comment (denyskon) [reply templates/shared/search/code/results.tmpl]: Turned out this template voodoo wasn't needed :laughing: Removed completely and replaced with a simple $.Link call

File: templates/shared/search/code/results.tmpl (hunk 15 above)
Comment (denyskon) [reply templates/shared/search/code/results.tmpl]: SImplified just to decide between .Repo or $.RepoID in the beginning

File: templates/repo/commits_table.tmpl (hunk 22 above)
Comment (denyskon) [reply templates/repo/commits_table.tmpl]: Done

**Comment by denyskon** [comment]
@silverwind @wxiaoguang I think I've addressed everything. Should be ready to merge.

**Comment by silverwind** [comment]
Doing one final style fix :)

**Comment by silverwind** [comment]
Fix done, it was the right border here not being blue:

Screenshot 2024-03-14 at 22 39 22

**Review by delvh — COMMENTED** [review]
Apparently I was too slow

File: templates/shared/search/code/results.tmpl (hunk 14 above)
Comment (denyskon) [reply templates/shared/search/code/results.tmpl]: > The parent template, or the Go code?

There is a `.Link` variable in every parent template that gives you the current relative URL as we always tass through the page object to that code search template, we can simply use it.

File: templates/admin/org/list.tmpl (hunk 19 above)
Comment (wxiaoguang) [reply templates/admin/org/list.tmpl:10]: I can answer your question. That's caused by an over-abstraction in old days.

At that time, `admin/base/search` was designed to be a "shared template" and was expected to be used everywhere. But over time, that design didn't fit more and more cases, so the "shared template" usages were replaced by dedicated code block one by one, then, here is the only last one place which uses the "shared template" nowadays.
